### ISSUE TITLE
Can accessible name be non-existent? Only whitespace?

### ISSUE THREAD

**Issue by Jym77** [act-rules/act-rules.github.io#976]
This stems from the discussions in https://github.com/act-rules/act-rules.github.io/pull/400#discussion_r321176440 and https://github.com/act-rules/act-rules.github.io/pull/400#discussion_r321201851

This is really several different questions but the way they interact with each other (and the ACT Rules) makes it worth tackling both at the same time…

Useful link: [Algorithm for computing accessible name](https://www.w3.org/TR/accname-1.1/#mapping_additional_nd_te).

Useful test case (using `role="link"` to allow name from content):
```html
<div role="link" id="A">A</div>
<div role="link" id="empty"></div>
<div role="link" id="1-space"> </div>
<div role="link" id="2-spaces">  </div>
<div role="link" id="newline">

</div>
<div role="link" id="br"><br /></div>
<div role="link" id="nbsp">&nbsp;</div>
<div role="link" id="A-nbsp-B">A &nbsp; B</div>
<div role="link" id="flat-string">A           B
  C
</div>
<div role="link" id="trim">  hello  </div>

<div aria-label="A" id="aria-A"></div>
<div aria-label="" id="aria-empty"></div>
<div aria-label=" " id="aria-1-space"></div>
<div aria-label="  " id="aria-2-spaces"></div>
<div aria-label="
" id="aria-newline"></div>
<div aria-label="&nbsp;" id="aria-nbsp"></div>
<div aria-label="A &nbsp; B" id="aria-A-nbsp-B"></div>
<div aria-label="A          B
    C" id="aria-flat-string"></div>
<div aria-label="  hello  " id="aria-trim"></div>
```

# Can accessible names be non-existent?
According to the algorithm, no. [Step 1](https://www.w3.org/TR/accname-1.1/#step1) set the `total accumulated text` to the empty string (`""`); [Steps 2](https://www.w3.org/TR/accname-1.1/#step2) only adds to it, never removes from or deletes it; and the end of the algorithm (no step number number) states:
> After all steps are completed, the total accumulated text is used as the accessible name or accessible description of the element that initiated the computation.

Therefore, as per the algorithm, the accessible name can be an empty string but always exists. Any formulation in the line of "something has an accessible name" is always true or should be specified as "something has a non-empty accessible name".

When testing with the code above and elements with id `empty` and `aria-empty`:
* Chrome gives them a name of `""` (empty string).
* Edge gives them no name (nothing is shown for the value of "Name", I do not know whether this means there is no value or whether his means an empty string that Edge decides not to display, especially since Edge does not display the double quotes for delimiting strings like Chrome or Firefox do).
* Firefox gives them a name of `null`. Which is plain wrong according to the algorithm.

# Can accessible name contain newline or several spaces in a row?
This is the point where the algorithm is not clear 😞 

On one hand, the Section [Name and Description](https://www.w3.org/TR/accname-1.1/#mapping_additional_nd) starts with:
> The output is a flat, unstructured string

and the Section [Accessible name and Description Copmutation](https://www.w3.org/TR/accname-1.1/#mapping_additional_nd_te) starts with (2nd paragraph):
> The purpose of the computation is to create a perceivable label or description for alternative presentations, in the form of a flat string of space separated textual tokens.

And flat string is defined as replacing newlines by spaces and merging consecutive spaces. So there should be no newline or consecutive spaces.

On the other hand, I do not see when in the algorithm is the string supposed to be flattened. Especially, [Step 2.G](https://www.w3.org/TR/accname-1.1/#step2G) (for text node) returns the textual content without flattening. My hypothesis here is that "textual content" implies flattening, but I fail to find a definition of "textual content" so far. Additionally, [Step 2C](https://www.w3.org/TR/accname-1.1/#step2C) says to return the value of the `aria-label` attribute without mentioning anything about flattening or textual content.
This gets really annoying when compared to [Step 2D](https://www.w3.org/TR/accname-1.1/#step2D) which handles `title` (and similar attributes) by _explicitly_ saying that this content should be flattened.

So, it seems that the accessible name must be a flat string but I do not really see which part of the algorithm enforces it…

In any case, for elements with id `flat-string` and `aria-flat-string`, both Chrome, Edge and Firefox are flattening the string and return a name of `"A B C"`; and for elements with id `trim` and `aria-trim`, the string is additionally trimmed of whitespace and the name is `"hello"` in all cases.

# Can the accessible name be only whitespace?
This is getting tricky… As for the previous case, nothing in the algorithm tells to trim whitespace. The only mention being in [Step 2C](https://www.w3.org/TR/accname-1.1/#step2C) where an `aria-label` that is empty when trimmed of whitespace is not processed. But otherwise, [Step 2C](https://www.w3.org/TR/accname-1.1/#step2C) does not require trimming the `aria-label` before adding it, and [Step 2G](https://www.w3.org/TR/accname-1.1/#step2G) does not require trimming text nodes before adding them (but again, maybe "textual content" means trimming but I can't find the definition).

Now, testing get fun… 
* As mentioned above for elements with id `trim` and `aria-trim`, the name is `"hello"` with Chrome, Edge and Firefox indicating that trimming did occur.
* For elements with id `1-space`, `2-spaces`, `newline`, `br`, `aria-1-space`, `aria-2-spaces` and `aria-newline`, the behaviour is the same as for `empty`/`aria-empty` (Chrome uses the empty string as name, Edge uses nothing and Firefox uses `null`). This indicates trimming and is coherent with the empty case.
* For element with id `nbsp`, Edge and Firefox keep the same behaviour (nothing or `null`) **but** Chrome gives a name of `" "` (this looks like a normal space according to my test (char code is 32), but there might be some other trick (as in copy/paste bad magic) turning a non breaking space into a normal one, so I don't guarantee what it actually is…).
* For element with id `aria-nbsp`, things start to get fun… Edge keeps giving nothing as name, but both Chrome and Firefox gives `" "` (again, not sure if it's space or non breaking space). This indicates that Firefox does **not** treat `aria-label` in the same way as text node (not the same kind of flattening is happening).

# Can accessible name contain several spaces in a row?
Algorithm description starts by saying that the result is a flat string, _i.e._ where consecutive spaces have been collapsed into a single one. So, in theory, no. This may of course depends on what exactly is called a space (ASCII whitespace or more?) And again, no step in the Algorithm indicates when this flattening should occur.

And tests are…
* Both for element with id `A-nbsp-B` and `aria-A-nbsp-B`, Edge gives a name of `"A B"` and both Chrome and Firefox give it a name of `"A   B"` (5 characters total (markdown messes this up 😞), the 3 in the middle look like normal spaces).

# How does that affect us?
The good point is that all the stuff about flattening the string doesn't really affect us. Whether the accessible name is `"A B"` or `"A     B"` (different number of whitespace) is usually not a concern for us.

The **very bad point** is about the non-existence and not only whitespace parts. It seems that an accessible name must always exist (as per the algorithm), even if Firefox disagrees. And concerning whitespace, all seem to at least agree on ASCII whitespace (they are trimmed). Other whitespace are more tricky.

So, taking a rule like [Button has accessible name](https://act-rules.github.io/rules/97a4e1), the (only) expectation states:
> Each target element has an accessible name that is not only whitespace.

Well, each element has an accessible name, and it cannot be only whitespace, so every applicable element should pass that rule.
Same goes for rule proposal #400 (from which this discussion started). Or any other rule checking that an accessible name exists and is not only whitespace.

# What to do?
The least to do is probably to replace all the "has an accessible name which is not only whitespace" by "has a non-empty accessible name". But we may probably want to investigate a bit more and decide what to do. Notably:
* What do we mean by "whitespace" (ASCII whitespace or also non breaking spaces)?
* How to handle browsers discrepancies (Firefox returning `null` instead of the empty string)?
* How to handle differences of handling between name from content (text nodes) and `aria-label` by Firefox?

Especially given that the algorithm for computing accessible name does not clearly answer these questions or explain what should be done in these corner cases.

Potentially affected rules:
```sh
$ grep -rl "has an \[accessible name" _rules/
_rules/auto-play-audio-has-control-mechanism-4c31df.md
_rules/button-accessible-name-97a4e1.md
_rules/form-control-accessible-name-e086e5.md
_rules/iframe-accessible-name-cae760.md
_rules/image-accessible-name-23a2a8.md
_rules/image-button-accessible-name-59796f.md
_rules/image-filename-as-accessible-name-9eb3f6.md
_rules/link-accessible-name-c487ae.md
```
(plus maybe some other using another formulation).

[edit: fix test case `A-nbsp-B` by adding regular spaces around the non breaking space]
[edit: replace spaces by hyphens in id]

**Comment by jkodu** [comment]
I think previously we had `has a non-empty accessible name` and because non-empty could not be properly defined, we went with `not only whitespace`.

It looks like we are coming to a full circle on this. 

@WilcoFiers has a thorough understanding of the accessible name computation algorithm. I would like to see his opinion on the computation logic revolving around whitespace.

**Comment by Jym77** [comment]
…and more fun stuff happens with `img` for which an empty `alt` attribute has a special meaning (saying the image is presentational)…

We already know that Chrome and Firefox are not respecting empty `alt` attribute on `img` elements (and put the image in the accessibility tree with role of `img` (Chrome) or `graphics` (Firefox): #692 

Let's play with these examples:
```html
<img src="https://picsum.photos/100" alt="A" id="A" /> 
<img src="https://picsum.photos/100" alt="" id="empty" />
<img src="https://picsum.photos/100" alt=" " id="1 space" />
<img src="https://picsum.photos/100" alt="  " id = "2 spaces" />
<img src="https://picsum.photos/100" alt="
" id="newline" />
<img src="https://picsum.photos/100" alt="&nbsp;" id="npsp" />
```

* Chrome consistently puts all of these in the accessibility tree with a role of `img`. It consistently give them an empty name, except for the non breaking space which is turned into a space. The handling of the `alt` attribute as accessible name is coherent with the rest of its behaviour, the handling of the empty `alt` attribute is incorrect according to specs, but that is somewhat unrelated.
* Edge correctly removes `empty` from the accessibility tree. `1 space`, `2 spaces`, `newline` and `nbsp` are in the accessibility tree without a name (or with a name which is empty or only spaces, it's hard to tell in Edge accessibility inspector). That is coherent with the rest of its behaviour for accessible names.
* Firefox consistently puts all of these in the accessibility tree with a role of `graphic` (which is super bad, but it's unrelated). Where it starts to be fun, is that `empty` has an empty accessible name, `1 space`, `2 spaces` and `nbsp` have an accessible name of `" "` (so not trimmed, but flattened) and `newline` has an accessible name of `"\n"` 😭 💥  That is completely not coherent with the rest of the behaviour for accessible name.

My gut feeling on the `img` part of this issue is that `<img alt="" />` should not be in the accessibility tree and `<img alt=" " />` (and variations) should be in with an empty accessible name.

**Comment by Jym77** [comment]
On a purely set theoretical side, I agree that an empty name is composed only of whitespace and thus should be rejected if testing "not only whitespace" (an empty name is also composed only of Chinese characters, for all it matters…)

I think I would rather have the rules testing "non empty" (`""`) and adding a note (and/or accessibility support) saying that whitespace is trimmed (or should be) and thus "only whitespace" is the same as empty. Or maybe keeping the "only whitespace" formulation but still adding a note (and/or accessibility support).

# Going through most of the "affected" rules
## Rules that are logically correct but using "only whitespace" instead of "empty":
* [Form control has accessible name](https://act-rules.github.io/rules/e086e5)
* [`iframe` element has accessible name](https://act-rules.github.io/rules/cae760)
* [Image button has accessible name](https://act-rules.github.io/rules/59796f)
* [Link has accessible name](https://act-rules.github.io/rules/c487ae)

These rules are technically OK. I'd rather see them with the small change + added note mentioned above.

## Image rule where `alt=""` is not the same as `alt=" "`
* [Image has accessible name](https://act-rules.github.io/rules/23a2a8)

I'm less sure what to do with this rule. The simple fix mentioned above is probably not enough and some more explanation should be added.

## Rules comparing accessible name with stuff where trimming and flattening should be taken into consideration
The first two of these are also hit by #968…
* [`iframe` elements with identical accessible names have equivalent purpose](https://act-rules.github.io/rules/4b1c6c)
* [Links with identical accessible names have equivalent purpose](https://act-rules.github.io/rules/b20e66)
* [Visible label is part of accessible name](https://act-rules.github.io/rules/2ee8b8)

These have no problem _per se_ but will probably benefit from improved examples to clarify how "identical" names is handling trimming and flattening of the names (are `" hello "` and `"hello"` identical names and do they need to be considered by these rules); or how trimming and flattening interacts with the inclusion of label into the name (trimming (and case-insensitivity) is already mentioned in the rule, flattening should also be mentioned).

## PR that are affected
In addition to the PR changing rules already mentioned above.
* #419 (Explicit SVG has accessible name).
Given the bloody mess that is accessible names for SVG, I haven't run test on them to see what is happening. In any case, this rule probably joins either the first case ("only whitespace" is the same as empty) or the second (adding the `alt=""` is decorative for `img` on top of it).
* #400 (Heading name is not only whitespace).
That one is a mess in its current state (which is why it sparkled this discussion…) given that it passes heading "without" accessible name (Passed Examples 1 and 2) but rejects heading with only whitespace/newline (Failed Examples 1, 2, and 3). In its current state, this rule is broken. I do not know why it was written as "empty is OK, whitespace is KO". Thus I have no real idea as to how to fix it.

# Questions we need to consider
* Are we happy with "non only whitespace" or do we want some other formulation? My answer is to use "non empty" and add note that "only whitespace" is (should be) the same as empty.
* Discrepancies between specs and browsers (Firefox giving a name of `null`). A classical issue. The classical solution seems to be to stick with specs (especially when they are already W3C recommendation as accessible name computation is) and add Accessibility Support mentioning deviant behaviours.
* Whitespace vs ASCII whitespace vs non breaking space vs … Chrome handles non breaking space differently than regular space. The specs are not clear on what should be done here. This may need more research (both in specs and browser testing) to understand what should happen and what is happening before we can make a decision.
* Incoherent handling of different sources of accessible name. Chrome and Firefox are performing different flattening of name from content and from `aria-label` (when interacting with non breaking space). Firefox is handling name from `alt` in yet another way. I haven't test other sources of names. Again, the specs are not super clear on what should be done but my gut feeling is that all sources of accessible name should be treated the same way.

**Comment by Jym77** [comment]
Mostly closed by #981.

Siteimprove to decide what to do about #400.
